packages/headless/src/api/search/search-api-client.ts

```diff
@@ -263,20 +265,34 @@ export class SearchAPIClient implements FacetSearchAPIClient {
     };
   }
 
-  private getAbortControllerInstanceIfAvailable(): AbortController | null {
-    // For nodejs environments only, we want to load the implementation of AbortController from node-abort-controller package.
-    // For browser environments, we need to make sure that we don't use AbortController as it might not be available (Locker Service in Salesforce)
-    // This is not something that can be polyfilled in a meaningful manner.
-    // This is a low level browser API after all, and only JS code inside a polyfill cannot actually cancel network requests done by the browser.
+  private getNewAbortController(
+    origin: SearchOrigin | 'unknown',
+    warnOnAbort: boolean
+  ) {
+    const newAbortController = createAbortController();
+    const oldAbortController = this.searchAbortControllers[origin];
+    if (newAbortController) {
+      this.searchAbortControllers[origin] = newAbortController;
+    }
+    if (oldAbortController) {
+      if (warnOnAbort) {
+        this.options.logger.warn('Cancelling current pending search query');
+      }
+      oldAbortController.abort();
+    }
+    return newAbortController;
+  }
 
-    if (typeof window === 'undefined') {
-      // eslint-disable-next-line @typescript-eslint/no-var-requires
-      const nodeAbort = require('node-abort-controller');
-      return new nodeAbort() as AbortController;
+  private clearAbortController(
+    origin: SearchOrigin | 'unknown',
+    controllerToClear: AbortController
+  ) {
+    if (
+      this.searchAbortControllers[origin]?.signal !== controllerToClear.signal
+    ) {
+      return;
     }
```

Comment:
Added this to prevent possible race conditions. I tried to write unit tests to verify the race condition, but they always failed, even with that fixed, and it took too long to dig into what was causing them to fail. I suspect the thing that caused them to fail was some weird behaviour I observed with Promises and mock timers in Jest, since stepping through the code executed in an order that didn't seem logical.